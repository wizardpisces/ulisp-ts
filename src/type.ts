export type Literal = string | number
export type Operator = string
export type FuncName = string
export type Expression = [Operator, Literal, Expression]

export type FuncDefinition = [FuncName, Literal[], ...Expression[]]
export type IfExpression = [Expression, Expression, Expression]

export type Token = string | number
export type Tokens = (Tokens | Token)[]