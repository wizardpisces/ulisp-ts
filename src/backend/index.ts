export {default as llvm} from './llvm'
export {default as x86} from './x86'