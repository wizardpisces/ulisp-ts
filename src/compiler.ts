import os from 'os';

type Arg = string | number
type Operator = string
type Args = [Operator, Arg, Args]
// type TupleType<T extends any[]> = {
//     [P in keyof T]: T[P]
// }
// type InstructionValue = TupleType<Args>

const SYSCALL_MAP = os.platform() === 'darwin' ? {
    'exit': '0x2000001',
} : {
        'exit': 60,
    };

let context = {
    assembly:'',
    push(code:string){
        context.assembly+=code+'\n';
    }
}

function emit(depth: number, code: string) {
    const indent = new Array(depth + 1).map(() => '').join('  ');

    // console.log(indent + code);
    context.push(indent + code);
}

function emit_prefix() {
    emit(1, '.global _main\n');

    emit(1, '.text\n');

    emit(0, 'plus:');
    emit(1, 'ADD RDI, RSI');
    emit(1, 'MOV RAX, RDI');
    emit(1, 'RET\n');

    emit(0, '_main:');
}

function emit_postfix() {
    emit(1, 'MOV RDI, RAX'); // Set exit arg
    emit(1, `MOV RAX, ${SYSCALL_MAP['exit']}`); // Set syscall number
    emit(1, 'SYSCALL');
}

function compile_argument(arg: Args[number], destination: string) {
    // If arg AST is a list, call compile_call on it
    if (Array.isArray(arg)) {
        compile_call(arg[0], (arg as Args).slice(1) as Args, destination);
        return;
    }

    // Else must be a literal number, store in destination register
    emit(1, `MOV ${destination}, ${arg}`);
}

const BUILTIN_FUNCTIONS: { [key: string]: string } = { '+': 'plus' };
const PARAM_REGISTERS:string[] = ['RDI', 'RSI', 'RDX'];

function compile_call(fun: string, args: Args, destination?:string) {
    // Save param registers to the stack
    args.forEach((_: Args[number], i: number) => emit(1, `PUSH ${PARAM_REGISTERS[i]}`));

    // Compile arguments and store in param registers
    args.forEach((arg: Args[number], i: number) => compile_argument(arg, PARAM_REGISTERS[i]));

    // Call function
    emit(1, `CALL ${BUILTIN_FUNCTIONS[fun] || fun}`);

    // Restore param registers from the stack
    args.forEach((_: Args[number], i:number) => emit(1, `POP ${PARAM_REGISTERS[args.length - i - 1]}`));

    // Move result into destination if provided
    if (destination) {
        emit(1, `MOV ${destination}, RAX`);
    }

    emit(0, ''); // For nice formatting
}

export function compile(ast:any) {
    emit_prefix();
    compile_call(ast[0], ast.slice(1));
    emit_postfix();
    return context
};