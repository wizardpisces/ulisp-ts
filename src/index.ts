export * from './parser'
export * from './type'