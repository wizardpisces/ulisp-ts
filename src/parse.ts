export type Token = string | number
export type Tokens = (Tokens | Token)[]

const WHITESPACE = [' ', '\n', '\t'];
export function parse(program: string): [Tokens, string] {
    const tokens: Tokens = [];
    let currentToken = '';

    for (let i = 0; i < program.length; i++) {
        const char = program.charAt(i);

        if (char === '(') {
            const [parsed, rest] = parse(program.substring(i + 1));
            tokens.push(parsed);
            program = rest;
            i = -1;
        } else if (char === ')') {
            if (currentToken.length) {
                tokens.push(+currentToken || currentToken);
            }

            return [tokens, program.substring(i + 1)];
        } else if (char === ';') {
            while (program.charAt(i) !== '\n') {
                i++;
            }
        } else if (WHITESPACE.includes(char)) {
            if (currentToken.length) {
                tokens.push(+currentToken || currentToken);
            }
            currentToken = '';
        } else {
            currentToken += char;
        }
    }

    return [tokens, ''];
};