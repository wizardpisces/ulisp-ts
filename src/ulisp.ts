import { parse } from './parse';
import { compile } from './compiler';

function main(args: any[]) {
    const script = args[2];
    const [ast] = parse(script);
    compile(ast[0]);
}

main(process.argv);