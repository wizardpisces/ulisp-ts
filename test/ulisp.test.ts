import { parse,Tokens } from '../src';
import {compile} from '../src/compiler'

describe('ulisp', () => {
  let program = '(+ 31 (+ 1 2))'
  it('parse', () => {
    let tokens: Tokens = [[["+", 31, ["+", 1, 2]]], ""]
    expect(JSON.stringify(parse(program))).toEqual(JSON.stringify(tokens));
  });
  it('compile',()=>{
    const script = program;
    // const result = `.global _main
    //   .text
    // plus:  ADD RDI, RSI  MOV RAX, RDI  RET
    // _main:  PUSH RDI  PUSH RSI  MOV RDI, 31  PUSH RDI  PUSH RSI  MOV RDI, 1  MOV RSI, 2  CALL plus  POP RSI  POP RDI  MOV RSI, RAX  CALL plus  POP RSI  POP RDI  MOV RDI, RAX  MOV RAX, 0x2000001  SYSCALL`
    const [ast] = parse(script);
    let context = compile(ast[0]);
    expect(context.assembly).toMatchSnapshot();
    console.log(context.assembly)
  })
});
