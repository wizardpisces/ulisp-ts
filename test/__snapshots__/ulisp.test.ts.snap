// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ulisp compile 1`] = `
"  .global _main

  .text

plus:
  ADD RDI, RSI
  MOV RAX, RDI
  RET

_main:
  PUSH RDI
  PUSH RSI
  MOV RDI, 31
  PUSH RDI
  PUSH RSI
  MOV RDI, 1
  MOV RSI, 2
  CALL plus
  POP RSI
  POP RDI
  MOV RSI, RAX

  CALL plus
  POP RSI
  POP RDI

  MOV RDI, RAX
  MOV RAX, 0x2000001
  SYSCALL
"
`;
